### PR TITLE
Add interface to IPC with management capabilities.

### PR DESCRIPTION
**Pull request by mamokarz**

### CODE REVIEW

File: src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+static az_result interface_manager_1_set_default_span_wrapper(
+    az_span model_in_span,
+    az_span* model_out_span)
+{
+  AZ_ULIB_TRY
+  {
+    // Unmarshalling JSON in model_in_span to interface_manager_model_in.
+    az_json_reader jr;
+    interface_manager_1_set_default_model_in interface_manager_model_in = { 0 };
```
Comment (mamokarz) [inline src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c:35]: >{ 0 } [](http://example.com/codeflow?start=74&length=5)

Clang in the gate didn't like this initialization.

File: inc/az_ulib_ipc_api.h [hunk 2]
```diff
@@ -454,6 +454,52 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ * @pre     \p package_name shall not be `NULL`.
+ * @pre     \p package_version shall not be `NULL`.
+ * @pre     \p interface_name shall not be `NULL`.
+ * @pre     \p interface_version shall not be `NULL`.
+ * @pre     \p capability_name shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the call.
+ *  @retval #AZ_OK                      If the method full name was split with success and can be
+ *                                      used.
+ *  @retval #AZ_ERROR_UNEXPECTED_CHAR   If there the method full name didn't fit the above rules.
```
Comment (mamokarz) [inline inc/az_ulib_ipc_api.h:492]: >there [](http://example.com/codeflow?start=43&length=5)

typo

File: src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c (hunk 1 above)
Comment (mamokarz) [reply src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c:35]: Disable -Wmissing-braces for now.

File: inc/az_ulib_interface_manager_1_model.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+#include "azure/core/_az_cfg_prefix.h"
+
+/*
+ * interface definition
+ */
+#define INTERFACE_MANAGER_1_INTERFACE_NAME "interface_manager"
+#define INTERFACE_MANAGER_1_INTERFACE_VERSION 1
+
+/*
+ * Define set_default command on interface_manager interface.
```
Comment (smcd253) [inline inc/az_ulib_interface_manager_1_model.h:27]: Does this comment describe all the names velow or just `INTERFACE_MANAGER_1_SET_DEFAULT_COMMAND_NAME`?

File: inc/az_ulib_interface_manager_1_model.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ */
+#define INTERFACE_MANAGER_1_SET_DEFAULT_COMMAND (az_ulib_capability_index)0
+#define INTERFACE_MANAGER_1_SET_DEFAULT_COMMAND_NAME "set_default"
+#define INTERFACE_MANAGER_1_SET_DEFAULT_INTERFACE_NAME "interface"
+#define INTERFACE_MANAGER_1_SET_DEFAULT_PACKAGE_NAME_NAME "package_name"
+#define INTERFACE_MANAGER_1_SET_DEFAULT_PACKAGE_VERSION_NAME "package_version"
+#define INTERFACE_MANAGER_1_SET_DEFAULT_INTERFACE_NAME_NAME "interface_name"
+#define INTERFACE_MANAGER_1_SET_DEFAULT_INTERFACE_VERSION_NAME "interface_version"
+typedef struct
```
Comment (smcd253) [inline inc/az_ulib_interface_manager_1_model.h:36]: if this is public, can you add a descriptor to this struct? Also add a newline above here.

File: inc/az_ulib_ipc_api.h [hunk 5]
```diff
@@ -454,6 +454,52 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
 
+/**
+ * @brief   Split the provided method full name.
+ *
+ * This function splits the provided method name between device, package, interface and capability
+ * using the following rules
+ *
+ * 1. The method full name shall be split by `.`.
+ * 2. The names and versions shall be extract from right to left.
+ * 3. The expected method full name shall looks like:
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:465]: nit: "... shall look like:"

File: inc/az_ulib_ipc_api.h [hunk 6]
```diff
@@ -454,6 +454,52 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ *      <device_name>.<package_name>.<package_version>.<interface_name>.<interface_version>.<capability_name>
+ * 4. The interface_name and interface_version are mandatory.
+ * 5. If a name is not provided, this function shall return #AZ_SPAN_EMPTY.
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:468]: can you specify "interface name" here?

File: inc/az_ulib_ipc_api.h [hunk 7]
```diff
@@ -454,6 +454,52 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ * 6. If a version is not provided, this function shall return #AZ_ULIB_VERSION_DEFAULT.
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:469]: same with version.

File: inc/az_ulib_ipc_api.h [hunk 8]
```diff
@@ -454,6 +454,52 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+ * @pre     IPC shall already be initialized.
+ * @pre     \p full_name shall not be #AZ_SPAN_EMPTY.
+ * @pre     \p device_name shall not be `NULL`.
+ * @pre     \p package_name shall not be `NULL`.
+ * @pre     \p package_version shall not be `NULL`.
+ * @pre     \p interface_name shall not be `NULL`.
+ * @pre     \p interface_version shall not be `NULL`.
+ * @pre     \p capability_name shall not be `NULL`.
+ *
+ * @return The #az_result with the result of the call.
```
Comment (smcd253) [inline inc/az_ulib_ipc_api.h:489]: You mention above in the rules that this function returns AZ_SPAN_EMPTY and AZ_ULIB_VERSION_DEFAULT.

File: src/az_ulib_ipc/az_ulib_ipc.c [hunk 9]
```diff
@@ -686,6 +710,133 @@ AZ_NODISCARD az_result az_ulib_ipc_call_with_str(
...
+  _az_PRECONDITION_NOT_NULL(capability_name);
+
+  AZ_ULIB_TRY
+  {
+    uint32_t package_version_temp;
+    int32_t interface_name_session;
+    uint32_t interface_version_temp;
+    bool has_capability_name;
+    int32_t count_session = 0;
+    az_span sessions[MAX_SESSIONS_IN_METHOD_FULL_NAME];
```
Comment (smcd253) [inline src/az_ulib_ipc/az_ulib_ipc.c:738]: The name "sessions" makes me think of network connections or maybe references to an object. Can you rename to something like "names", "fields", or "descriptors"?

File: tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c [hunk 10]
```diff
@@ -2643,6 +2874,184 @@ static void az_ulib_ipc_get_table_succeed(void** state)
   /// cleanup
 }
 
+/* The az_ulib_ipc_split_method_name shall split the method full name returning device, package,
```
Comment (smcd253) [inline tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c:2877]: shouldn't these tests be with the tests ending on line 937?

**Review by smcd253 — CHANGES_REQUESTED** [review]
See comments and address gate failures

File: inc/az_ulib_interface_manager_1_model.h (hunk 4 above)
Comment (mamokarz) [reply inc/az_ulib_interface_manager_1_model.h:36]: I agree with both comments here, I'll improve the documentation.

File: inc/az_ulib_interface_manager_1_model.h (hunk 3 above)
Comment (mamokarz) [reply inc/az_ulib_interface_manager_1_model.h:27]: It is for all. But I'll improve it.

File: inc/az_ulib_ipc_api.h (hunk 6 above)
Comment (mamokarz) [reply inc/az_ulib_ipc_api.h:468]: It is not the interface, it is for other names, I'll specify it.

File: inc/az_ulib_ipc_api.h (hunk 7 above)
Comment (mamokarz) [reply inc/az_ulib_ipc_api.h:469]: same

File: inc/az_ulib_ipc_api.h (hunk 8 above)
Comment (mamokarz) [reply inc/az_ulib_ipc_api.h:489]: Rule 4, interface_name and interface_version are mandatory.

File: src/az_ulib_ipc/az_ulib_ipc.c (hunk 9 above)
Comment (mamokarz) [reply src/az_ulib_ipc/az_ulib_ipc.c:738]: I like fields.

File: tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c (hunk 10 above)
Comment (mamokarz) [reply tests/tests_ut/az_ulib_ipc_ut/az_ulib_ipc_ut.c:2877]: No, those are the Precondition tests in the UT. They are inside of `AZ_NO_PRECONDITION_CHECKING`. After that we have the tests not related to the precondition.

File: src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c (hunk 1 above)
Comment (mamokarz) [reply src/az_ulib_ipc/az_ulib_ipc_interface_manager_interface.c:35]: Didn't work.

**Review by smcd253 — APPROVED** [review]
lgtm

**Pull request merged**